File: libs/power-log-parser/src/lib/parsers/linked-entity-parser.ts
import { GameTag } from '@firestone-hs/reference-data';
import { ActionParser } from '../action-parser';
import { GameEventProvider, GameEventHelper } from '../game-event';
import { FullEntity, Node, NodeType, ShowEntity, TagChange } from '../models';
import { GameState } from '../state/game-state';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

export class LinkedEntityParser implements ActionParser {
	readonly ParserName = 'LinkedEntityParser';

	private GameState: GameState;
	private ParserState: ParserState;
	private StateFacade: StateFacade;

	constructor(parserState: ParserState, facade: StateFacade) {
		this.ParserState = parserState;
		this.GameState = parserState.GameState;
		this.StateFacade = facade;
	}

	AppliesOnNewNode(node: Node, stateType: StateType): boolean {
		return (
			stateType === StateType.PowerTaskList &&
			node.Type === NodeType.TagChange &&
			(node.Object as TagChange).Name === (GameTag.LINKED_ENTITY as number)
		);
	}

	AppliesOnCloseNode(node: Node, stateType: StateType): boolean {
		if (stateType !== StateType.PowerTaskList) {
			return false;
		}
		// ShowEntity / FullEntity may embed LINKED_ENTITY (reconnect / mid-game Discover snapshots).
		if (node.Type === NodeType.ShowEntity) {
			return (node.Object as ShowEntity).GetTag(GameTag.LINKED_ENTITY) > 0;
		}
		if (node.Type === NodeType.FullEntity) {
			return (node.Object as FullEntity).GetTag(GameTag.LINKED_ENTITY) > 0;
		}
		return false;
	}

	CreateGameEventProviderFromNew(node: Node): GameEventProvider[] | null {
		const tagChange = node.Object as TagChange;
		const entity = this.GameState.CurrentEntities.get(tagChange.Entity)!;
		const cardId = entity.CardId;
		if (!this.GameState.CurrentEntities.has(tagChange.Value)) {
			return null;
		}

		const linkedEntity = this.GameState.CurrentEntities.get(tagChange.Value)!;
		if (linkedEntity?.Id === tagChange.Entity) {
			return null;
		}

		const controllerId = entity.GetEffectiveController();
		return [
			GameEventProvider.Create(
				tagChange.TimeStamp,
				'LINKED_ENTITY',
				GameEventHelper.CreateProvider('LINKED_ENTITY', cardId, controllerId, entity.Id, this.StateFacade, {
					LinkedEntityId: tagChange.Value,
					LinkedEntityControllerId: linkedEntity.GetEffectiveController(),
					LinkedEntityZone: linkedEntity.GetZone(),
					LinkedEntityCost: linkedEntity.GetCost(),
				}),
				true,
				node,
			),
		];
	}

	CreateGameEventProviderFromClose(node: Node): GameEventProvider[] | null {
		const source = node.Object as ShowEntity | FullEntity;
		const linkedEntityId = source.GetTag(GameTag.LINKED_ENTITY);
		if (!this.GameState.CurrentEntities.has(linkedEntityId)) {
			return null;
		}

		const linkedEntity = this.GameState.CurrentEntities.get(linkedEntityId)!;
		if (linkedEntity.Id === source.Entity) {
			return null;
		}

		// Same as COPIED_FROM: FullEntity Creating closes before CurrentEntities registration.
		const cardId = source.CardId;
		const controllerId = source.GetEffectiveController();
		// Opponent ShowEntity/FullEntity often embeds LINKED_ENTITY on dredge/discover previews.
		// Emitting LINKED_ENTITY here writes the revealed cardId onto the linked deck row and
		// bypasses CopiedFrom's dredge obfuscation (info leak). Keep TagChange LINKED as-is;
		// only emit close events for the local player's own links (e.g. Tracking Discover).
		if (controllerId !== this.StateFacade.LocalPlayer?.PlayerId) {
			return null;
		}
		return [
			GameEventProvider.Create(
				source.TimeStamp,
				'LINKED_ENTITY',
				GameEventHelper.CreateProvider('LINKED_ENTITY', cardId, controllerId, source.Entity, this.StateFacade, {
					LinkedEntityId: linkedEntityId,
					LinkedEntityControllerId: linkedEntity.GetEffectiveController(),
					LinkedEntityZone: linkedEntity.GetZone(),
					LinkedEntityCost: linkedEntity.GetCost(),
				}),
				true,
				node,
			),
		];
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/LinkedEntityParser.cs
using HearthstoneReplays.Parser;
using HearthstoneReplays.Parser.ReplayData;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using System.Linq;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using System.Collections.Generic;
using static HearthstoneReplays.Events.CardIds;

namespace HearthstoneReplays.Events.Parsers
{
    public class LinkedEntityParser : ActionParser
    {
        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade StateFacade { get; set; }

        public LinkedEntityParser(ParserState ParserState, StateFacade facade)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.StateFacade = facade;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            return stateType == StateType.PowerTaskList
                && node.Type == typeof(TagChange)
                && (node.Object as TagChange).Name == (int)GameTag.LINKED_ENTITY;
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            return false;
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            var tagChange = node.Object as TagChange;
            var entity = GameState.CurrentEntities[tagChange.Entity];
            var cardId = entity.CardId;
            // That can happen when starting to spectate a BG game partway through
            if (!GameState.CurrentEntities.ContainsKey(tagChange.Value))
            {
                return null;
            }

            // This seems to happen when playing ETC and the sideboard options are shown
            var linkedEntity = GameState.CurrentEntities[tagChange.Value];
            if (linkedEntity?.Id == tagChange.Entity)
            {
                return null;
            }

            // In the case of Spy-o-matic, the active player discovers a card, and so the new entities are 
            // attached to the active player. We in fact want to have the controllerId be the controller 
            // of the original card, so that we can properly transfer information to it
            // However, the card copy (CARD_REVEALED) is usually attached to the same player as the one 
            // from the LINKED_ENTITY, so we have to keep the same controller.
            // We'll add the LinkedEntityControllerId attribute instead to make the link
            var controllerId = entity.GetEffectiveController();
            //var gameState = GameEvent.BuildGameState(ParserState, StateFacade, GameState, tagChange, null);
            return new List<GameEventProvider> { GameEventProvider.Create(
                tagChange.TimeStamp,
                "LINKED_ENTITY",
                GameEvent.CreateProvider(
                    "LINKED_ENTITY",
                    cardId,
                    controllerId,
                    entity.Id,
                    StateFacade,
                    //gameState, 
                    new {
                        LinkedEntityId = tagChange.Value,
                        LinkedEntityControllerId = linkedEntity.GetEffectiveController(),
                        LinkedEntityZone = linkedEntity.GetZone(),
                        LinkedEntityCost = linkedEntity.GetCost(),
                    }),
                true,
                node) };
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            return null;
        }
    }
}
